fix host memory alloc and pinning on OpenPower systems

psychocoderHPC opened the pull request:
The alignment requirements to pin host memory on OpenPower systems are not fulfilled.

We tried to fix this issue already in the past in #873. Unfortunately, the issue still exists and was reproducible with PIConGPU and cuda-memcheck. After applying this fix the error is solved.

~~This PR introduces a new boost dependency to the header only lib `boost::interprocess` :-(.
I thought about fixing it with macros guards but we do not know if we will run on ARM+CUDA again into this issue or if the system is running with huge pages we will definitely have this issue again, therefore I decided to use boost.~~

@j-stephan Can we discuss in the next developer meeting if we create with this fix a bug fix release 0.9.1, else we need to switch with PIConGPU to the current alpaka dev for 1.0.

bernhardmgruber commented on include/alpaka/mem/alloc/AllocCpuAligned.hpp:
```diff
@@ -37,16 +39,16 @@ namespace alpaka
-                // to 4kib.
-                // @todo evaluate requirements when the HIP ecosystem is more stable
-                constexpr size_t minAlignement = std::max<size_t>(TAlignment::value, 4096);
+                // On most x86 systems the page size is 64KiB and on OpenPower 65KiB.
+                // Page size can be tested on the terminal with: `getconf PAGE_SIZE`
```
It's still 4KiB on my machine :)
```
bgruber@graviola:~$ getconf PAGE_SIZE
4096
```
Is that a typo?

psychocoderHPC replied:
yes it schould be 4kib

psychocoderHPC commented:
mhh looks like the boost part is not header only. :-(

j-stephan commented:
I'm not sure I like Boost.Interprocess. This seems like a small-enough change to implement this ourselves (through POSIX and Win32-API).

j-stephan commented on include/alpaka/dev/cpu/SysInfo.hpp:
```diff
@@ -123,6 +123,26 @@ namespace alpaka
+            //! \return Pagesize in bytes used by the system.
+            inline size_t getPageSize()
+            {
+#if BOOST_OS_WINDOWS || BOOST_OS_CYGWIN
```
Wow, we support Cygwin?

psychocoderHPC replied:
not really but since it was in the original code pathes in I keeped it

psychocoderHPC committed:
add `getPageSize()` function

Add function to querry the system page size.

psychocoderHPC committed:
fix alignment requirements to pin host memory

The alignment requirements to pin host memory on OpenPower systems are not fulfilled.
This is a follow up of #873.

psychocoderHPC commented:
@j-stephan I implemented the `getPageSize()` method and used it for memory alignment and for other functions where we already used the page size for getting free memory.

bernhardmgruber commented on include/alpaka/dev/cpu/SysInfo.hpp:
```diff
@@ -123,6 +123,26 @@ namespace alpaka
+                GetSystemInfo(&si);
+                return si.dwPageSize;
+#elif BOOST_OS_UNIX || BOOST_OS_MACOS
+#    if defined(_SC_PAGESIZE)
+                return static_cast<std::size_t>(sysconf(_SC_PAGESIZE));
+#    else
+                return = static_cast<size_t>(getpagesize());
+#    endif
```
Consider adding a comment that `getpagesize` is deprecated and that's why we have two versions here.

j-stephan commented on include/alpaka/dev/cpu/SysInfo.hpp:
```diff
@@ -123,6 +123,26 @@ namespace alpaka
+                SYSTEM_INFO si;
+                GetSystemInfo(&si);
+                return si.dwPageSize;
+#elif BOOST_OS_UNIX || BOOST_OS_MACOS
```
Shouldn't `BOOST_OS_UNIX` also work for macOS?

psychocoderHPC replied:
I do not know I have no MAC and other code snippeds in alpaka use the same combination.

j-stephan commented on include/alpaka/dev/cpu/SysInfo.hpp:
```diff
@@ -123,6 +123,26 @@ namespace alpaka
+                return si.dwPageSize;
+#elif BOOST_OS_UNIX || BOOST_OS_MACOS
+#    if defined(_SC_PAGESIZE)
+                return static_cast<std::size_t>(sysconf(_SC_PAGESIZE));
+#    else
+                return = static_cast<size_t>(getpagesize());
```
Are there still systems around which don't use `_SC_PAGESIZE`? `getpagesize()` was deprecated [before 1997](https://pubs.opengroup.org/onlinepubs/7908799/xsh/getpagesize.html) (see the copyright header).

psychocoderHPC replied:
not sure the mange page in linux says that with sysconf  you are more portable but getpagesize is still available in linux.

j-stephan commented on include/alpaka/dev/cpu/SysInfo.hpp:
```diff
@@ -156,21 +176,17 @@ namespace alpaka
+#    elif defined(_SC_PHYS_PAGES) // Linux, FreeBSD, OpenBSD, Solaris.
+                return static_cast<std::size_t>(sysconf(_SC_PHYS_PAGES)) * getPageSize();
 
-#    elif defined(_SC_PHYS_PAGES) && defined(_SC_PAGE_SIZE) // Legacy.
-                return static_cast<std::size_t>(sysconf(_SC_PHYS_PAGES))
-                    * static_cast<std::size_t>(sysconf(_SC_PAGE_SIZE));
+#    elif defined(_SC_PHYS_PAGES) // Legacy.
+                return static_cast<std::size_t>(sysconf(_SC_PHYS_PAGES)) * getPageSize();
```
Now we have two `#elif defined(_SC_PHYS_PAGES)`. I think we can merge those.

j-stephan replied:
@psychocoderHPC You missed this review comment. We have two branches doing the same thing here.

psychocoderHPC replied:
you are right :-(

j-stephan commented on include/alpaka/mem/alloc/AllocCpuAligned.hpp:
```diff
@@ -37,16 +38,15 @@ namespace alpaka
-                // to 4kib.
-                // @todo evaluate requirements when the HIP ecosystem is more stable
-                constexpr size_t minAlignement = std::max<size_t>(TAlignment::value, 4096);
+                // On most x86 systems the page size is 64KiB and on OpenPower 65KiB.
```
I agree with Bernhard here. On x86 it is probably 4KiB.

j-stephan reviewed (changes requested):
Thanks! A few nitpicks:

psychocoderHPC commented:
I updated the PR.
